src/algebra/non_unital_alg_hom.lean

```diff
@@ -0,0 +1,171 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Morphisms of non-unital algebras
+
+This file defines morphisms between two modules, each of which also carries a multiplication that is
+compatible with the additive structure. The multiplications are not assumed to be associative or
+unital, or even compatible with the scalar multiplication. In a typical application, the
+multiplications and scalar multiplications will satisfy compatibility conditions making them into
+algebras (albeit possibly non-associative and/or non-unital) but such conditions are not required to
+make this definition.
+
+## Main definitions
+
+  * `non_unital_alg_hom`
+  * `alg_hom.non_unital_alg_hom.has_coe`
+
+## Tags
+
+non-unital, algebra, morphism
+-/
+
+universes u v w w₁ w₂ w₃
+
+variables (R : Type u) (A : Type v) (B : Type w) (C : Type w₁)
+
+set_option old_structure_cmd true
+
+/-- A morphism respecting addition, multiplication, and scalar multiplication. When these arise from
+algebra structures, this is the same as a not-necessarily-unital morphism of algebras. -/
+structure non_unital_alg_hom [semiring R]
+  [non_unital_non_assoc_semiring A] [module R A]
+  [non_unital_non_assoc_semiring B] [module R B]
+  extends A →ₗ[R] B, mul_hom A B
```

Comment:
In theory you could relax this to
```suggestion
structure non_unital_alg_hom [monoid R]
  [non_unital_non_assoc_semiring A] [distrib_mul_action R A]
  [non_unital_non_assoc_semiring B] [distrib_mul_action R B]
  extends A →+[R] B, mul_hom A B
```

which then lets you express things like `non_unital_alg_hom (units R) A B`

Reply:
Hmm, this is _probably_ a good idea. I've just added a commit which does it anyway so let's see if it builds etc.